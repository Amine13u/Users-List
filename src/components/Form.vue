<template>
  <v-row justify="center">
    <v-dialog v-model="dialog" persistent max-width="600px">
      <template v-slot:activator="{ on, attrs }">
        <v-btn
          icon
          v-bind="attrs"
          v-on="on"
          class="mt-3"
          @click="toggleDialog(true)"
        >
          <v-icon>mdi-plus</v-icon>
        </v-btn>
      </template>
      <v-card>
        <v-card-title>
          <span class="text-h5">User Profile</span>
        </v-card-title>
        <v-card-text>
          <v-container>
            <v-row>
              <v-col cols="12" sm="6" md="4">
                <v-text-field
                  label="first name*"
                  required
                  v-model="form.firstName"
                ></v-text-field>
              </v-col>
              <v-col cols="12" sm="6" md="4">
                <v-text-field
                  label="last name*"
                  required
                  v-model="form.lastName"
                ></v-text-field>
              </v-col>

              <v-col cols="12">
                <v-text-field
                  label="Email*"
                  required
                  v-model="form.email"
                ></v-text-field>
              </v-col>
            </v-row>
          </v-container>
          <small>*indicates required field</small>
        </v-card-text>
        <v-card-actions>
          <v-spacer></v-spacer>
          <v-btn color="blue darken-1" text @click="toggleDialog(false)">
            Close
          </v-btn>
          <v-btn color="blue darken-1" text @click="submitForm()">
            Save
          </v-btn>
        </v-card-actions>
      </v-card>
    </v-dialog>
  </v-row>
</template>

<script>
import { mapState } from "vuex";
export default {
  data: () => ({
    form: { firstName: "", lastName: "", email: "" },
  }),
  methods: {
    submitForm() {
      if (Object.values(this.form).includes(""))
        return alert("One field is missing");

      this.$store.commit(this.form.id ? "updateUser" : "addUser", this.form);
      this.form = { firstName: "", lastName: "", email: "" };
      this.toggleDialog(false);
    },
    toggleDialog(status) {
      this.$store.commit("toggleDialog", { status });
    },
  },
  watch: {
    editForm(next) {
      // if next value is truthy
      // we will initialize the form with the user data
      if (next) {
        this.form = { ...next };
      }
    },
  },
  computed: mapState({
    dialog: (state) => state.dialog,
    editForm: (state) => state.editedUser,
  }),
};
</script>
