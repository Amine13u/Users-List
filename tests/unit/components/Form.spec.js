import { shallowMount, createLocalVue } from "@vue/test-utils";
import Form from "@/components/Form.vue";
// import Vue from "vue";
import Vuex from "vuex";
// import Vuetify from "vuetify";

const localVue = createLocalVue();
//localVue.use(Vuex);

// const store = new Vuex.Store({});

describe("Form", () => {
  test("snapshot", () => {
    let $store = {
      state: {
        search: "",
        dialog: false,
        editedUser: {},
        users: [
          {
            id: 1,
            firstName: "Amine",
            lastName: "Mannai",
            email: "amine@gmail.com",
          },
          {
            id: 2,
            firstName: "Karim",
            lastName: "Gharbi",
            email: "karim@gmail.com",
          },
          {
            id: 3,
            firstName: "John",
            lastName: "Doe",
            email: "john@gmail.com",
          },
        ],
      },
    };

    const wrapper = shallowMount(Form, {
      mocks: { $store },
    });
    expect(wrapper).toMatchSnapshot();
  });
});
