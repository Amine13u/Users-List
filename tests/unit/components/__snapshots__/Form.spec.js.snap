// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form snapshot 1`] = `
<v-row-stub tag="div" justify="center">
  <v-dialog-stub opendelay="0" closedelay="0" contentclass="" maxwidth="600px" origin="center center" persistent="true" retainfocus="true" transition="dialog-transition" width="auto">
    <v-card-stub loaderheight="4" tag="div">
      <v-card-title-stub><span class="text-h5">User Profile</span></v-card-title-stub>
      <v-card-text-stub>
        <v-container-stub tag="div">
          <v-row-stub tag="div">
            <v-col-stub cols="12" sm="6" md="4" tag="div">
              <v-text-field-stub errorcount="1" errormessages="" messages="" rules="" successmessages="" backgroundcolor="" label="first name*" loaderheight="2" clearicon="$clear" type="text" required=""></v-text-field-stub>
            </v-col-stub>
            <v-col-stub cols="12" sm="6" md="4" tag="div">
              <v-text-field-stub errorcount="1" errormessages="" messages="" rules="" successmessages="" backgroundcolor="" label="last name*" loaderheight="2" clearicon="$clear" type="text" required=""></v-text-field-stub>
            </v-col-stub>
            <v-col-stub cols="12" tag="div">
              <v-text-field-stub errorcount="1" errormessages="" messages="" rules="" successmessages="" backgroundcolor="" label="Email*" loaderheight="2" clearicon="$clear" type="text" required=""></v-text-field-stub>
            </v-col-stub>
          </v-row-stub>
        </v-container-stub> <small>*indicates required field</small>
      </v-card-text-stub>
      <v-card-actions-stub>
        <v-spacer-stub></v-spacer-stub>
        <v-btn-stub color="blue darken-1" tag="button" activeclass="" text="true" type="button">
          Close
        </v-btn-stub>
        <v-btn-stub color="blue darken-1" tag="button" activeclass="" text="true" type="button">
          Save
        </v-btn-stub>
      </v-card-actions-stub>
    </v-card-stub>
  </v-dialog-stub>
</v-row-stub>
`;
